File: C++/fractals.cpp
#include "fractals.hpp"
#include <cstdlib>
#include <iostream>
#include <sstream>
#include <iomanip>

SDL_Window* window = NULL;
SDL_Renderer* renderer = NULL;
SDL_Event event;

int main()
{
  SDL_Init(SDL_INIT_VIDEO);
  window = SDL_CreateWindow("Mandelbrot", SDL_WINDOWPOS_UNDEFINED,
      SDL_WINDOWPOS_UNDEFINED, WIDTH, HEIGHT, SDL_WINDOW_SHOWN);
  renderer = SDL_CreateRenderer(window, -1, SDL_RENDERER_ACCELERATED);

  bool quit = false;

  init();
  
  int frame = 0;
  while (!quit)
  {
    while (SDL_PollEvent(&event) > 0) {
      if (event.type == SDL_QUIT)
        quit = true;
      if (event.type == SDL_MOUSEBUTTONDOWN && event.button.button == 1) {
        float nWidthHalf = (MAXX - MINX) / 4;
        float nHeightHalf = (MAXY - MINY) / 4;
        float nCenterX = double(event.button.x) / WIDTH  * (MAXX-MINX) + MINX;
        float nCenterY = double(event.button.y) / HEIGHT * (MAXY-MINY) + MINY;
        MINX = nCenterX - nWidthHalf;
        MAXX = nCenterX + nWidthHalf;
        MINY = nCenterY - nHeightHalf;
        MAXY = nCenterY + nHeightHalf;

        MAX_ITERATIONS += 5;

        break;
      }
    }

    SDL_SetRenderDrawColor(renderer, 0, 0, 0, 255);
    SDL_RenderClear(renderer);

    calculate_frame();

    for(int x = 0; x < WIDTH; x++) {
      for(int y = 0; y < HEIGHT; y++) {
        auto i = iters[x][y];
        SDL_SetRenderDrawColor(renderer, R[i], G[i], B[i], 255);
        SDL_RenderDrawPoint(renderer, x, y);
      }
    }

    // SDL_RenderPresent(renderer);
    if (frame < FRAMES) {
      const Uint32 format = SDL_PIXELFORMAT_ARGB8888;
      SDL_Surface* pScreenShot = SDL_CreateRGBSurfaceWithFormat(0, WIDTH, HEIGHT, 32, format);

      if(pScreenShot)
      {
        SDL_RenderReadPixels(renderer, NULL, format, pScreenShot->pixels, pScreenShot->pitch);

        ostringstream str;
        str << "/tmp/" << setfill('0') << setw(4) << frame << ".bmp";
        SDL_SaveBMP(pScreenShot, str.str().c_str());
        ++frame;

        SDL_FreeSurface(pScreenShot);
      } 
    }
    else
    {
      quit = true;
    }
  }
}

File: Nim/fractals.nim
import sdl2, sdl2/gfx
import complex

discard sdl2.init(INIT_EVERYTHING)

const
    MAX_ITERATIONS = 50

    MIN_X = -2
    MAX_X = 1

    MIN_Y = -1
    MAX_Y = 1

    WIDTH = 1200
    HEIGHT = 800

    NUM_THREADS = 8

discard sdl2.init(INIT_VIDEO)

var
    iters: array[0..WIDTH, array[0..HEIGHT, uint8]]

    event = sdl2.defaultEvent

    window = createWindow("Mandelbrot", 100, 100, WIDTH, HEIGHT, SDL_WINDOW_SHOWN)
    renderer = createRenderer(window, -1, RENDERER_ACCELERATED)

    quit = false
    invalid = true

    fpsman: FpsManager

fpsman.init
renderer.setDrawColor(0, 0, 0, 255)
renderer.clear()
renderer.present()

proc calcDot(x, y: int): uint8 = 
    var c : Complex = (toFloat(x) / WIDTH  * (MAX_X - MIN_X) + MIN_X,
                       toFloat(y) / HEIGHT * (MAX_Y - MIN_Y) + MIN_Y)
    var z = c
    while result < MAX_ITERATIONS:
        if abs(z) >= 4:
            return
        z = z * z + c
        result += 1


while not quit:
    while pollEvent(event):
        if event.kind == QuitEvent:
            quit = true
            break
    if invalid:
        for x in 0..<WIDTH:
            for y in 0..<HEIGHT:
                iters[x][y] = calcDot(x, y)

        renderer.setDrawColor(0, 0, 0, 255)
        renderer.clear()

        for x in 0..<WIDTH:
            for y in 0..<HEIGHT:
                renderer.setDrawColor(uint8(255 * float(iters[x][y]) / MAX_ITERATIONS),
                                      0,
                                      uint8(255 - (255 / MAX_ITERATIONS) * float(iters[x][y])))
                renderer.drawPoint(cint(x), cint(y))
    
        renderer.present()
        invalid = false
    fpsman.delay

destroy renderer
destroy window